qringbuf.go

```diff
@@ -1,3 +1,168 @@
+// Package qringbuf provides a concurrency-friendly zero-copy abstraction of
+// io.ReadAtLeast() over a pre-allocated ring-buffer, populated asynchronously
+// by a standalone goroutine. It is primarily designed for processing a series
+// of arbitrary streams each comprised of variable-length records.
+//
+// Specifically an object of this package makes the following guarantees:
+//  • Memory is allocated only at construction time, never during streaming
+//  • StartFill() spawns off a single goroutine (collector) which terminates when:
+//    ◦ It reaches readLimit, if one was supplied
+//    ◦ It receives any error from the wrapped reader (including io.EOF)
+//    ◦ It had a chance to evaluate a signal received from StopFill()
+//      (most io.Read() operations are not cancelable, and may block forever)
+//  • Every call to NextRegion() blocks until it can return:
+//    ◦ A *Region object representing a contiguous slice of at least MinRegion bytes
+//    ◦ A smaller "trailing" *Region and an io.EOF or ErrCollectorStopped
+//    ◦ A nil-Region and any other error
+//  • *Region.Bytes() is always a slice of the underlying buffer, no data copying takes place
+//  • Data backing a *Region is guaranteed to remain intact provided:
+//    ◦ NextRegion() has not been called again allowing writes into the buffer
+//    ◦ *Region.Reserve() was invoked, which blocks writes until a *Region.Release()
+//
+// Examples
+//
+// In code the basic usage looks roughly like this:
+//
+//  qrb, initErr := qringbuf.NewFromReader( *someReader, qringbuf.Config{ … } )
+//  …
+//  readLimit = int64( stopReadingAfterThatManyBytes )
+//  startErr := qrb.StartFill( readLimit )
+//  …
+//  var available, processed int
+//  for {
+//      reg, streamErr := qrb.NextRegion( available - processed )
+//      if reg == nil {
```

Comment:
Usually in go I'd expect to check `if streamErr != nil`, is there a reason to instead check `if reg == nil`?

Reply:
TLDR: this did end up making me change code, but the can of worms is strong with that one: https://github.com/ipfs/go-qringbuf/pull/6/commits/4d65af211de

Now the long version: This is how I wrote go at first too, and then got legit burned in some unrelated code, realizing **checking for error first/only is incorrect**. From https://golang.org/pkg/io/#Reader :

> When Read encounters an error or end-of-file condition after successfully reading n > 0 bytes, it returns the number of bytes read. ... Callers **should always process the n > 0 bytes returned before considering the error err**. Doing so correctly handles I/O errors that happen after reading some bytes and also both of the allowed EOF behaviors. 

In the example you are asking about, consider that `io.EOF` will start showing up the moment the collector reaches it ( and shuts down ). One will still cycle over the remaining, probably very large region, potentially multiple times requesting less and less of the remaining data: each of these calls will indicate `by the way, nothing extra is coming anymore`, but you are not in an actual "stop work" error condition.

On the other hand I did it wrong anyway: as it was, if the collector "eats" some data, and then the underlying stream errors: the data is now locked in the buffer without a way to get it out. So reworked as https://github.com/ipfs/go-qringbuf/pull/6/commits/4d65af211de

Reply:
Ok makes sense - I think the important thing is to make clear any deviation from idiomatic patterns and the example comments & explanation you added address that 👍